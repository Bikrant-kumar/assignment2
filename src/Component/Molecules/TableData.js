import React, {useState} from 'react';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import AddCircleOutlineOutlinedIcon from '@material-ui/icons/AddCircleOutlineOutlined';
import IconButton from '@material-ui/core/IconButton';
import AddItem from '../Atoms/AddItem';
import Collapse from '@material-ui/core/Collapse';
import { useDispatch } from 'react-redux';
import { useSelector } from 'react-redux';
import SendIcon from '@material-ui/icons/Send';
import DeleteIcon from '@material-ui/icons/Delete';
import Grid from '@material-ui/core/Grid';
import {deleteRow,deleteAllItem,addRow} from '../../Store/Action/ListAction';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Label from '../Atoms/Label';
import RemoveCircleOutlineIcon from '@material-ui/icons/RemoveCircleOutline';

const useStyles = makeStyles((theme) => ({
  paper: {
    marginLeft: "8%",
    marginRight: "8%",
    marginTop: theme.spacing(12),
    background: 'linear-gradient(60deg, #2196F3 30%, #21CBF3 90%)',
    boxShadow: '0 3px 5px 2px rgba(33, 203, 243, .3)',
  },
  CAlign:{
    textAlign:"center"
  }
}));

function TableData({table}) {
  const [open, setOpen] = useState(false);
  const [menu,setMenu] = useState([]);
  const dispatch = useDispatch();
  const items  = useSelector(state=>state.item);
  const getMenuName = (value)=>{
    setMenu(value);
    }
  const classes = useStyles();

  return (
          <Grid className={classes.paper}>
            <Grid className={classes.CAlign}> <Label text={table.Name} /></Grid>     
            <TableContainer component={Paper}>
            <Table size="small" aria-label="a dense table">
            <TableHead>
            <TableRow>
            <TableCell > 
              <IconButton color="primary" aria-label="upload picture" component="span" onClick={() => setOpen(!open)}> 
                {open ? <RemoveCircleOutlineIcon/> : <AddCircleOutlineOutlinedIcon/>}
              </IconButton>
              </TableCell>
              <TableCell>
              <Collapse in={open} timeout="auto" unmountOnExit>
                <AddItem getMenu={getMenuName} name="Menu" Details={items}/>
                <IconButton color="primary" aria-label="upload picture" component="span" onClick={()=> dispatch(addRow(table,menu)) && setOpen(!open)}> 
                  <SendIcon/> 
                </IconButton>
              </Collapse>
              </TableCell>
            <TableCell align="right">S No.</TableCell>
            <TableCell align="right">Name</TableCell>
            <TableCell align="right">Unit</TableCell>
            <TableCell align="right">Sum</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {table.item.map((row,index) => (
            <TableRow key={index} >
              <TableCell  >
              <IconButton color="primary" onClick={()=>dispatch(deleteRow(table,row.id,row.Cost)) && setMenu([])}> 
              <DeleteIcon />
              </IconButton>
                </TableCell>
                <TableCell></TableCell>
               <TableCell align="right">{index+1}</TableCell> 
              <TableCell align="right">{row.Name}</TableCell>
              <TableCell align="right">{row.Serving}</TableCell>
              <TableCell align="right">{row.Cost}</TableCell>
            </TableRow>
          ))}
          <TableRow>
              <TableCell > 
              {open ? <Button variant="contained" disabled >
                  Check Out
                </Button>:
                <Button color="primary" variant="outlined" onClick={()=>dispatch(deleteAllItem(table)) && setMenu([])} >
                  Check Out
                </Button>}
              </TableCell>
              <TableCell > </TableCell>
              <TableCell align="right"></TableCell>
              <TableCell align="right"></TableCell>
              <TableCell align="right"> Total </TableCell> 
              <TableCell align="right">{table.totalCost}</TableCell>
          </TableRow>
           </TableBody>
      </Table>
    </TableContainer>
    </Grid>
    )
}

export default TableData