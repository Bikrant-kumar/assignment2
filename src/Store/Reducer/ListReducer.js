export const initState = {    
   
};

export function itemReducer (state = initState, action) {
    switch (action.type) {
        case "DELETE_ITEM":
            const newItem = state.item.filter((item) => item.id !== action.id);
            return{
                ...state,
                item: newItem
            };

        case "DELETE_ROW":
            const newTable= action.table.item.splice(action.table.item.indexOf(action.rowId),1)
                action.table.totalCost = action.table.totalCost - action.rowCost;
                action.table.totalItem= action.table.totalItem- 1;
            return {
                ...state,
                ...newTable
            };
        
        case "ADD_ROW":
            const newRowInTable = action.table.item.push(action.menu)
            action.table.totalCost = action.table.totalCost + action.menu.Cost;
            action.table.totalItem= action.table.totalItem + 1;
            return {
                ...state,
                ...newRowInTable
            };

        case "ADD_ITEM":
            const items = [action.item, ...state.item]
            return {
                ...state,
                item: items,
            };
        case "ADD_REVIEW":
            const review = [action.review,...state.reviews]
            return{
                ...state,
                reviews:review
            };
        
        case "GET_STATE":
            return {  
                ...action.data,
            }

        case "DELETE_ALL_ITEM":
            const newTable1 = action.table.item.splice(0,action.table.item.length)
            action.table.totalCost = 0;
            action.table.totalItem = 0;
            return {
                ...state,
                ...newTable1
            }
            
        default:
            return state;    
    }
};

 export default itemReducer;