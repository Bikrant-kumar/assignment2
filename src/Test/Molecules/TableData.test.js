import React from 'react';
import { screen, render } from '@testing-library/react';
import TableData from '../../Component/Molecules/TableData';
import ListReducer from '../../Store/Reducer/ListReducer';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import MockData from '../MockData';
import { createStore } from 'redux';

describe('TableData',()=> {

    test('TableData Render',() => {
  
      const table={
        "id": "table-2",
        "Name": "Table2",
        "totalItem": 0,
        "totalCost": 0,
        "item": []
      };      
          render(
            <Provider store={createStore(ListReducer, MockData)}>
          <TableData table={table}/>
          </Provider>);
          expect(screen.getByText("Table2")).toBeInTheDocument();
       });
  
       test('test TableData with snapshot',()=>{
         const table={
        "id": "table-2",
        "Name": "Table2",
        "totalItem": 0,
        "totalCost": 0,
        "item": [
            {
                "id": "item-27",
                "Name": "Chenna Poda",
                "Cost": 150,
                "Serving":1,
                "Photo":"https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcRAv5Iol4J1UYRJ9sb5mqRV47D-v37ymZXe_A&usqp=CAU"
              }
        ]
      };  
        const component= renderer.create(
          <Provider store={createStore(ListReducer, MockData)}><TableData table = {table}/> </Provider>).toJSON();
        expect(component).toMatchSnapshot();
       });
  
  })
   