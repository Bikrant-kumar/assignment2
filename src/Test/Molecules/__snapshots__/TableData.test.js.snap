// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableData test TableData with snapshot 1`] = `
Array [
  <div
    className="MuiGrid-root makeStyles-paper-4"
  >
    <div
      className="MuiGrid-root makeStyles-CAlign-5"
    >
       
      <div
        className="MuiBox-root MuiBox-root-6"
      >
        Table2
      </div>
    </div>
    <div
      className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <table
        aria-label="a dense table"
        className="MuiTable-root"
        role={null}
      >
        <thead
          className="MuiTableHead-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root MuiTableRow-head"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
              scope="col"
            >
              <span
                aria-disabled={false}
                aria-label="upload picture"
                className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                role="button"
                tabIndex={0}
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                    />
                  </svg>
                </span>
              </span>
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
              scope="col"
            />
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
              scope="col"
            >
              S No.
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
              scope="col"
            >
              Name
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
              scope="col"
            >
              Unit
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
              scope="col"
            >
              Sum
            </th>
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
            >
              <button
                className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
              </button>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
            />
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall"
            >
              1
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall"
            >
              Chenna Poda
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall"
            >
              1
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall"
            >
              150
            </td>
          </tr>
          <tr
            className="MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiButton-label"
                >
                  Check Out
                </span>
              </button>
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
            >
               
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall"
            />
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall"
            />
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall"
            >
               Total 
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall"
            >
              0
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  " ",
]
`;
